compiler/rustc_data_structures/src/sync.rs

```diff
@@ -146,7 +146,7 @@ cfg_if! {
             t.into_iter()
         }
 
-        pub fn par_for_each_in<T: IntoIterator>(t: T, for_each: impl Fn(T::Item) + Sync + Send) {
+        pub fn par_for_each_in<T: IntoIterator>(t: T, mut for_each: impl FnMut(T::Item) + Sync + Send) {
```

Comment:
Wouldn't this break the parallel compiler? The requirement is relaxed and then if we switch to parallel rustc it would fail to compile.

Reply:
It doesn't break the compiler from actual testing. 

This function signature change is only required in non-parallel mode, and [the one in parallel mode](https://github.com/rust-lang/rust/blob/master/compiler/rustc_data_structures/src/sync.rs#L338) does not need to be changed. I think this is because the internal implementation of `IntoParallelIterator` does some kind of encapsulation of this.

Reply:
I don't think we can do that, if we run it in parallel then the function would be called from multiple threads thus requiring it to be called without having exclusive access.

Reply:
That makes sense. So I think we can keep the `Fn` in the signature in parallel mode, as it doesn't cause compilation errors

Reply:
I revisited the implementation of both patterns. The place to call this interface is in [collect.rs](https://github.com/rust-lang/rust/blob/master/compiler/rustc_monomorphize/src/collector.rs#L349). Because the MTRef type is not a `&mut` reference [in parallel mode](https://github.com/rust-lang/rust/blob/master/compiler/rustc_data_structures/src/sync.rs#L268), the `Fn` type can be used without causing a compilation error.